File: pt6/1276.number-of-burgers-with-no-waste-of-ingredients.go
package main

func numOfBurgers(tomatoSlices int, cheeseSlices int) []int {
	extraTomatoes := tomatoSlices - 2*cheeseSlices
	if extraTomatoes < 0 || extraTomatoes%2 != 0 {
		return []int{}
	}

	jumbo := extraTomatoes / 2
	small := cheeseSlices - jumbo
	if small < 0 {
		return []int{}
	}

	return []int{jumbo, small}
}

/*
Explanation

Let jumbo=x and small=y. Then x+y=cheeseSlices and 4x+2y=tomatoSlices.
Subtract 2*(x+y) from the tomato equation and get 2x = tomatoSlices -
2*cheeseSlices. That gives jumbo directly; small is the remaining cheese
count.

Both counts must be nonnegative integers.

Edge cases: too few tomatoes, odd extraTomatoes, or jumbo greater than total
cheese all mean impossible.

Time complexity: O(1).
Space complexity: O(1).
*/

File: pt6/1276.number-of-burgers-with-no-waste-of-ingredients.java
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

class Solution {
    public List<Integer> numOfBurgers(int tomatoSlices, int cheeseSlices) {
        int extraTomatoes = tomatoSlices - 2 * cheeseSlices;
        if (extraTomatoes < 0 || extraTomatoes % 2 != 0) {
            return Collections.emptyList();
        }

        int jumbo = extraTomatoes / 2;
        int small = cheeseSlices - jumbo;
        if (small < 0) {
            return Collections.emptyList();
        }

        return Arrays.asList(jumbo, small);
    }
}

/*
Explanation

Let jumbo = x and small = y. Then x + y = cheeseSlices and
4x + 2y = tomatoSlices. Subtracting 2 * cheeseSlices from tomatoSlices leaves
2x, so x = (tomatoSlices - 2 * cheeseSlices) / 2.

After computing jumbo, small is the remaining cheese count. Both counts must be
nonnegative integers.

Edge cases: too few tomatoes; odd extra tomato count; jumbo greater than total
cheese.

Time complexity: O(1).
Space complexity: O(1).
*/
